lib/zendesk_apps_support/finders.rb

```diff
@@ -0,0 +1,32 @@
+module ZendeskAppsSupport
+  module Finders
+    class RecordNotFound < StandardError
+      def initialize(msg)
+        super(msg)
+      end
+    end
+
+    def find_by(arg)
+      all.find(&filter_by_arg(arg))
+    end
+
+    def find_by!(arg)
+      found = find_by(arg)
+      fail RecordNotFound.new("Unable to find #{self.name} with #{arg.inspect}") if found.nil?
+      found
+    end
+
+    def where(arg)
+      all.select(&filter_by_arg(arg))
+    end
+
+    private
+
+    def filter_by_arg(arg)
+      fail('More than one key-value pair found') if arg.size > 1
```

Comment:
I found this a little hard to follow. I think it might be easier if we explicitly state which values you can `find_by` and maybe have a better error message. I don't think this code is better than yours but maybe we can find a way to improve. I find converting hashes to array always a little odd, maybe just me.

```ruby
class Location
  can_be_found_by :code

  def self.can_be_found_by(attribute)
    @finders ||= []
    @finders << attribute
  end
  def self.finders
    @finders
  end
  def finders
    class.finders
  end
  def find_by(arg)
    raise "#{self.class} can only be queried using a single condition." if arg.keys.size != 1
    finder = arg.keys[0]
    raise "#{self.class} can only be queried by #{finders}" unless finders.include? finder
    value = arg.delete(finder)
    all.find {|l| l.public_send(finder) == value }
  end
end
```